### PR TITLE
Zcash AddressDecoder

### PR DESCRIPTION
**Pull request by tok-kkk**
Currently the `AddressEncodeDecoder` for zcash will not return an error when trying to decode an address on different network, for example:
```golang
addrEncodeDecoder := zcash.NewAddressEncodeDecoder(&zcash.MainNetParams)
addr := multichain.Address("t28Tc2BUTHifXthexsohy89umGdqMWLSUqw") // Testnet address 
_, err := addrEncodeDecoder.DecodeAddress(addr)
Expect(err).To(HaveOccurred())
```
I expect this to return an error when trying to decode a testnet address using a mainnet `AddressEncodeDecoder`.
All other bitcoin-like chains are working fine and I expect zcash to have the same behaviour.  

The `param` we passed to initialize the `AddressEncodeDecoder` is not used when encoding/decoding.  And the 
 `parsePrefix` function checks all possible networks. I changed that to compare only with the param of the `AddressEncodeDecoder`.

### CODE REVIEW

**Review by loongy — APPROVED** [review]

**Review by jazg — APPROVED** [review]
Looks good to me as well. Will just wait for CI to pass before merging.

**Review by roynalnaruto — APPROVED** [review]
Looks good! I added another commit to clean up some redundant code, and to propagate better errors. Will merge when the build passes!

**Review by roynalnaruto — APPROVED** [review]

**Pull request merged**
